### DISCUSSION TITLE
Workaround for dynamic `namePath` for Form?

### DISCUSSION THREAD

**Discussion by bombillazo** [ant-design/ant-design#45041] [Q&A] [answered]
Hello, we want to "watch" the value of a for.item that is configurable via props, including the path name. Is there any way to get around the limitation of needing a static path for useWatch with forms?

```bash
Warning: `useWatch` is not support dynamic `namePath`. Please provide static instead. 
```

**Comment by zombieJ** [comment, marked as answer]
You can do watch on the top level and wrap it:
```tsx
const useMyWatch = (namePath, form) => {
  const root = useWatch('');
  return get(root, namePath);
};
```

**Reply by crazyair** [reply]
```tsc
const data = useWatch('');
const dynamicName=data.dynamicName 
```

**Reply by bombillazo** [reply]
> You can do watch on the top level and wrap it:
> 
> ```tsx
> const useMyWatch = (namePath, form) => {
>   const root = useWatch('');
>   return get(root, namePath);
> };
> ```

are you using lodash `get` here?

**Reply by bombillazo** [reply]
Also I tried the solution but it prevents rerenders on the value change since the hook is looking at the root only seems the internal data changes do not trigger renders...

**Reply by lilinhao1054** [reply]
> Also I tried the solution but it prevents rerenders on the value change since the hook is looking at the root only seems the internal data changes do not trigger renders...

if you have looked `rc field form` source code, you'll find that when the controls(wrapped in form, like input/select/custom...) value change, `updateValue` will be called, and then `notifyWatch` will be called, it will call all the watch callbacks in `watchList`,

when we use `useWatch`, it will call `registerWatch`, and push the callback into `watchList`.

which means every time the control value change, all the watchCb will be called, include the callback registered by `useWatch`, but it does not conclude endless re-render, because only the value we watched change, `setState` will be called, and cause re-render

```js
  private registerWatch: InternalHooks['registerWatch'] = callback => {
    this.watchList.push(callback);

    return () => {
      this.watchList = this.watchList.filter(fn => fn !== callback);
    };
  };

  private notifyWatch = (namePath: InternalNamePath[] = []) => {
    // No need to cost perf when nothing need to watch
    if (this.watchList.length) {
      const values = this.getFieldsValue();
      const allValues = this.getFieldsValue(true);

      this.watchList.forEach(callback => {
        callback(values, allValues, namePath);
      });
    }
  };


  private updateValue = (name: NamePath, value: StoreValue) => {
    const namePath = getNamePath(name);
    const prevStore = this.store;
    this.updateStore(setValue(this.store, namePath, value));

    this.notifyObservers(prevStore, [namePath], {
      type: 'valueUpdate',
      source: 'internal',
    });
    this.notifyWatch([namePath]);
    ...

  // in useWatch
  const getWatchValue = (values: any, allValues: any) => {
      const watchValue = options.preserve ? allValues : values;
      return typeof dependencies === 'function'
        ? dependencies(watchValue)
        : getValue(watchValue, namePathRef.current);
    };
  
    const cancelRegister = registerWatch((values, allValues) => {
      const newValue = getWatchValue(values, allValues);
      const nextValueStr = stringify(newValue);
  
      // Compare stringify in case it's nest object
      if (valueStrRef.current !== nextValueStr) {
        valueStrRef.current = nextValueStr;
        setValue(newValue);
      }
    });
```
about the solution, root will be undefined in my version `antd 5.22`, the value will be values[''] = undefined, you can use `useWatch([])` or `useWatch(values => values)` instead
```js
const useMyWatch = (namePath, form) => {
  const root = useWatch('');
  return get(root, namePath);
};
```
**BUT WARNING**: the solution, will cause endless re-render, any value in form change, `setState` in `useWatch` will be called, because of the comparison based on stringfied object

Let's try to meet halfway on that.
```js
type keyType = keyof any;
export const useDynamicWatch = (staticPath: keyType[], dynamicPath: keyType[], form?: FormInstance) => {
  const root = Form.useWatch(staticPath, form);
  return get(root, dynamicPath);
};
```
only re-render in staticPath's value change, not the whole form

**Comment by Henry-G-H-Huang** [comment]
May i know why the form.useWatch can't accept the Dynamic path?

**Reply by crazyair** [reply]
https://github.com/react-component/field-form/pull/758

**Reply by Henry-G-H-Huang** [reply]
> [react-component/field-form#758](https://github.com/react-component/field-form/pull/758)

想问下这个升级在ant design的新版本已经带进来吗 具体是哪个版本

**Reply by crazyair** [reply]
https://github.com/ant-design/ant-design/blob/next/package.json#L124 antd v6

**Reply by Henry-G-H-Huang** [reply]
BTW，if watch the var by selector, will it avoid the rc-form bug?
```
const A = Form.useWatch(value => _.get(value, ['xxxxx', indexA, 'yyyyy', indexB, 'zzzzzz']))
```

**Reply by Henry-G-H-Huang** [reply]
试了一下好像不行 应该是有闭包的问题 导致一直读的都是旧的index. 还是要等v6 想问问v6 计划什么时候正式release

**Reply by crazyair** [reply]
https://github.com/ant-design/ant-design/issues/54173 是这个问题，就是这个问题上面修复的
